### DISCUSSION TITLE
Is there any way to add or edit subquestion and answer options?

### DISCUSSION THREAD

**Discussion by microplasticos** [edgarrmondragon/citric#1131] [Q&A] [answered]
I was working quite well with Citric, I managed to create surveys, groups, and import questions, but I hit a wall in my development when I realized that the API doesn't incorporate a method that allows me to edit subquestions and answer options. Is this a limitation of the API or of Citric? Is there anything I can contribute to resolve this limitation? Thanks.

**Comment by edgarrmondragon** [comment]
Hi @microplasticos! Thanks for the feedback. Maybe https://github.com/edgarrmondragon/citric/discussions/971#discussioncomment-6979702 is helpful, have you seen tried something like that?

**Reply by edgarrmondragon** [reply]
More concretely,

```python
client.set_question_properties(123, subquestions={<qid>: ...})
```

**Reply by microplasticos** [reply]
Thank you, I will try and I'll let you know. In the meantime, I am trying out a way to edit and add subquestions and response options to the lsq file before importing it.

**Reply by edgarrmondragon** [reply]
@microplasticos I just gave the `client.set_question_properties` approach a try and it's indeed _not_ possible to set sub-questions that way.

That method doesn't really do anything with a `subquestions` parameter, so you can't add sub-questions to a survey question. You also can't create questions and set the parent question ID to make them sub-questions.

You're right in attempting to use an lsq file to create your question and sub-questions. I've added use case to #1119.

**Reply by microplasticos** [reply]
Yeah, I spent a whole afternoon trying the same thing until I came to the same conclusion. It's a major bummer not being able to add options and subquestions using `client.set_question_properties`. I'll let you know as soon as I get some successful results generating an lsq file.
Thank you!

**Reply by microplasticos** [reply]
In the end, I gave up, and I'll try to create a survey through database queries. It's a pity that the LimeSurvey API is so limited in that regard. Thanks for your help!

**Reply by edgarrmondragon** [reply]
A pity indeed! I submitted a feature request, let's see what the maintainers think: https://bugs.limesurvey.org/view.php?id=19574

**Reply by microplasticos** [reply]
Man, thanks for submitting this as a feature request. You're the GOAT.

**Reply by RicWainwright1** [reply]
I am commenting here to track any changes.

**Reply by edgarrmondragon** [reply]
https://github.com/edgarrmondragon/citric/issues/1186 might be a solution.

**Comment by edgarrmondragon** [comment]
Hello there! 👋

I just v2.0.0 of the library, which includes a new method to update answer options in a question.

Here's a quick example:

```python
from citric.rest import RESTClient

client = RESTClient(
    "https://example.com/rest/v1",
    "your_username",
    "your_password",
)

survey_id = 142423

# Get survey details
survey = client.get_survey(survey_id)
print(survey)

# Inspect question 247 to find answer IDs
for question in survey["questionGroups"][0]["questions"]:
    if answers := question.get("answers"):
        print("Answers for question", question["qid"])
        for answer in answers:
            print(f"  Answer ID: {answer['aid']}, Code: {answer['code']}")

# Define patch operations
operations = [
    {
        "entity": "answer",
        "op": "update",
        # Question ID
        "id": 247,
        "props": [
            {
                # Answer ID
                "aid": 31,
                "code": "AO01",
                "sortOrder": 1,
                "assessmentValue": 0,
                "scaleId": 1,
                "l10ns": {
                    "de": {"answer": "ANTW1 scale 1", "language": "de"},
                    "en": {"answer": "ANS1 scale 1", "language": "en"},
                },
            },
            {
                # Answer ID
                "aid": 32,
                "code": "AO02",
                "sortOrder": 1,
                "assessmentValue": 0,
                "scaleId": 1,
                "l10ns": {
                    "de": {"answer": "ANTW2 scale 1", "language": "de"},
                    "en": {"answer": "ANS2 scale 1", "language": "en"},
                },
            },
            {
                # Answer ID
                "aid": 33,
                "code": "AO03",
                "sortOrder": 1,
                "assessmentValue": 0,
                "scaleId": 1,
                "l10ns": {
                    "de": {"answer": "ANTW3 scale 1", "language": "de"},
                    "en": {"answer": "ANS3 scale 1", "language": "en"},
                },
            },
        ],
    }
]


result = client.patch_survey(survey_id, operations)
print(f"Operations applied: {result['operationsApplied']}")  # Operations applied: 1
```

It's also [available in the docs](https://citric.readthedocs.io/en/latest/how-to.html#update-question-answers-using-the-rest-api).

I would love to hear your feedback!

N.B. I'd like to add a more high-level API for this, but the LimeSurvey REST API seems to still be in-flux, so I'm waiting for it to stabilize before doing so.

**Comment by edgarrmondragon** [comment, marked as answer]
https://github.com/edgarrmondragon/citric/pull/1676 will let you do this

```python
mc_question = Question(
    title="Q02",
    type="M",
    l10ns={"en": QuestionL10n(question="Which of the following do you use?")},
    subquestions=[
        Question(
            title="SQ001",
            type="T",
            l10ns={"en": QuestionL10n(question="Python")},
        ),
        Question(
            title="SQ002",
            type="T",
            l10ns={"en": QuestionL10n(question="R")},
        ),
        Question(
            title="SQ003",
            type="T",
            l10ns={"en": QuestionL10n(question="Julia")},
        ),
    ],
)

client.import_question(mc_question.to_lsq(), survey_id, group_id)
```

This will be available in version 2.1.0.
